### PR TITLE
Switched to a single labeled metric for the time in q in zipper. Start of zipper redesign.

### PR DESCRIPTION
**Pull request by grzkv**
Note: This PR includes unfinished refactoring. Some rough edges are to be expected.

## What issue is this change attempting to solve?
* Fixes time in queue for zipper stats.
* Refactoring/redesign of the zipper.
  * Removed parts of the global state
  * Partially removed expvars
  * Restructured modules
  * Misc cleanups

### CODE REVIEW

File: pkg/backend/net/net.go [hunk 1]
```diff
@@ -238,17 +243,25 @@ func (b Backend) do(trace types.Trace, req *http.Request) (string, []byte, error
...
-func (b Backend) call(ctx context.Context, trace types.Trace, u *url.URL) (string, []byte, error) {
+func (b Backend) call(ctx context.Context, trace types.Trace, u *url.URL, request string) (string, []byte, error) {
 	ctx, cancel := b.setTimeout(ctx)
 	defer cancel()
 
-	t0 := time.Now()
+	var t *prometheus.Timer
+	if b.qHist != nil { // TODO: remove condition when capi is merged with zipper
+		t = prometheus.NewTimer(b.qHist.WithLabelValues(request))
+		t.ObserveDuration()
```
Comment (emadolsky) [inline pkg/backend/net/net.go:253]: I think we shouldn't observe duration here.

File: pkg/backend/net/net.go [hunk 2]
```diff
@@ -238,17 +243,25 @@ func (b Backend) do(trace types.Trace, req *http.Request) (string, []byte, error
...
+	}
+
 	err := b.enter(ctx)
-	trace.AddLimiter(t0)
 	if err != nil {
 		return "", nil, err
 	}
 
+	if t != nil {
```
Comment (emadolsky) [inline pkg/backend/net/net.go:261]: If I'm not mistaken, having this after the error check, we'll miss the times in queues which lead to error (e.g. timeout). I think we want to observe those times as well.

File: pkg/backend/net/net.go [hunk 3]
```diff
@@ -370,7 +383,7 @@ func (b Backend) Info(ctx context.Context, request types.InfoRequest) ([]types.I
 	u = carbonapiV2InfoEncoder(u, metric)
 	request.Trace.AddMarshal(t0)
 
-	_, resp, err := b.call(ctx, request.Trace, u)
+	_, resp, err := b.call(ctx, request.Trace, u, "info")
```
Comment (emadolsky) [inline pkg/backend/net/net.go:386]: Why don't we keep `trace.AddLimiter` and use the time stored there after `b.call` to observe the time on histogram?
This way, we omit this `request string` argument which looks a bit weird IMHO.

File: pkg/backend/net/net.go (hunk 1 above)
Comment (grzkv) [reply pkg/backend/net/net.go:253]: Yep, we shouldn't.

File: pkg/backend/net/net.go (hunk 2 above)
Comment (grzkv) [reply pkg/backend/net/net.go:261]: I don't think we should. This is intentional but debatable.

File: pkg/backend/net/net.go (hunk 3 above)
Comment (grzkv) [reply pkg/backend/net/net.go:386]: `trace` will go away, no need to keep it. Also, it averages stats which is wrong.

File: pkg/backend/net/net.go (hunk 1 above)
Comment (grzkv) [reply pkg/backend/net/net.go]: Pushed. Thanks for finding it!

**Review by emadolsky — APPROVED** [review]
We discussed the PR with Roman, and as there are a lot of things that will still change in the process, we decided to continue, and do not bother ATM.

**Pull request merged**